File: examples/dart-hello-world/tool/connect_example.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:grpc/grpc.dart';
import 'package:holons/holons.dart' as holons;

Future<File> sdkEchoServer() async {
  final path = File(
    '${Directory.current.path}/../../sdk/dart-holons/bin/echo-server',
  ).absolute;
  if (!await path.exists()) {
    throw StateError('echo-server not found at ${path.path}');
  }
  return path;
}

Future<void> writeEchoHolon(Directory root, File binaryPath) async {
  final holonDir = Directory('${root.path}/holons/echo-server');
  await holonDir.create(recursive: true);
  final holonYaml = File('${holonDir.path}/holon.yaml');
  await holonYaml.writeAsString('''
uuid: "echo-server-connect-example"
given_name: Echo
family_name: Server
motto: Reply precisely.
composer: "connect-example"
kind: service
build:
  runner: dart
  main: bin/echo-server
artifacts:
  binary: "${binaryPath.path}"
''');
}

Future<void> main() async {
  final root = await Directory.systemTemp.createTemp(
    'dart-holons-connect-example-',
  );
  final previousCwd = Directory.current;

  try {
    await writeEchoHolon(root, await sdkEchoServer());
    Directory.current = root.path;

    final channel = await holons.connect('echo-server');
    try {
      final method = ClientMethod<String, String>(
        '/echo.v1.Echo/Ping',
        (request) => utf8.encode(request),
        (payload) => utf8.decode(payload),
      );

      final call = channel.createCall(
        method,
        Stream<String>.value('{"message":"hello-from-dart"}'),
        CallOptions(timeout: const Duration(seconds: 5)),
      );

      stdout.writeln(await call.response.single);
    } finally {
      await holons.disconnect(channel);
    }
  } finally {
    Directory.current = previousCwd.path;
    if (await root.exists()) {
      await root.delete(recursive: true);
    }
  }
}

File: examples/hello-world/rust-hello-world/src/connect_example.rs
use bytes::{Buf, BufMut};
use std::env;
use std::error::Error;
use std::fs;
use std::path::PathBuf;
use tempfile::TempDir;
use tonic::client::Grpc;
use tonic::codec::{Codec, DecodeBuf, Decoder, EncodeBuf, Encoder};
use tonic::codegen::http::uri::PathAndQuery;
use tonic::{Request, Status};

#[derive(Clone, Default)]
struct StringCodec;

#[derive(Clone, Default)]
struct StringEncoder;

#[derive(Clone, Default)]
struct StringDecoder;

impl Codec for StringCodec {
    type Encode = String;
    type Decode = String;
    type Encoder = StringEncoder;
    type Decoder = StringDecoder;

    fn encoder(&mut self) -> Self::Encoder {
        StringEncoder
    }

    fn decoder(&mut self) -> Self::Decoder {
        StringDecoder
    }
}

impl Encoder for StringEncoder {
    type Item = String;
    type Error = Status;

    fn encode(&mut self, item: Self::Item, dst: &mut EncodeBuf<'_>) -> Result<(), Self::Error> {
        dst.reserve(item.len());
        dst.put_slice(item.as_bytes());
        Ok(())
    }
}

impl Decoder for StringDecoder {
    type Item = String;
    type Error = Status;

    fn decode(&mut self, src: &mut DecodeBuf<'_>) -> Result<Option<Self::Item>, Self::Error> {
        let bytes = src.copy_to_bytes(src.remaining());
        String::from_utf8(bytes.to_vec())
            .map(Some)
            .map_err(|err| Status::internal(err.to_string()))
    }
}

struct CurrentDirGuard {
    previous: PathBuf,
}

impl CurrentDirGuard {
    fn enter(path: &std::path::Path) -> Result<Self, Box<dyn Error>> {
        let previous = env::current_dir()?;
        env::set_current_dir(path)?;
        Ok(Self { previous })
    }
}

impl Drop for CurrentDirGuard {
    fn drop(&mut self) {
        let _ = env::set_current_dir(&self.previous);
    }
}

fn sdk_echo_server() -> Result<PathBuf, Box<dyn Error>> {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../../sdk/rust-holons/bin/echo-server")
        .canonicalize()?;
    if !path.is_file() {
        return Err(format!("echo-server not found at {}", path.display()).into());
    }
    Ok(path)
}

fn write_echo_holon(
    root: &std::path::Path,
    binary_path: &std::path::Path,
) -> Result<(), Box<dyn Error>> {
    let holon_dir = root.join("holons/echo-server");
    fs::create_dir_all(&holon_dir)?;
    fs::write(
        holon_dir.join("holon.yaml"),
        format!(
            concat!(
                "uuid: \"echo-server-connect-example\"\n",
                "given_name: Echo\n",
                "family_name: Server\n",
                "motto: Reply precisely.\n",
                "composer: \"connect-example\"\n",
                "kind: service\n",
                "build:\n",
                "  runner: rust\n",
                "  main: bin/echo-server\n",
                "artifacts:\n",
                "  binary: \"{}\"\n"
            ),
            binary_path.display()
        ),
    )?;
    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let temp_root = TempDir::new()?;
    let echo_server = sdk_echo_server()?;
    write_echo_holon(temp_root.path(), &echo_server)?;
    let _cwd = CurrentDirGuard::enter(temp_root.path())?;

    let channel = holons::connect::connect("echo-server").await?;
    let mut grpc = Grpc::new(channel.clone());
    grpc.ready().await?;
    let response = grpc
        .unary(
            Request::new("{\"message\":\"hello-from-rust\"}".to_string()),
            PathAndQuery::from_static("/echo.v1.Echo/Ping"),
            StringCodec,
        )
        .await?
        .into_inner();

    println!("{response}");
    holons::connect::disconnect(channel).await?;
    Ok(())
}
